code/modules/paperwork/paper.dm

```diff
@@ -248,147 +438,292 @@
+					return TRUE
+
+			var/obj/item/holding = user.get_active_held_item()
+			// Use a clipboard's pen, if applicable
+			if(istype(loc, /obj/item/clipboard))
+				var/obj/item/clipboard/clipboard = loc
+				// This is just so you can still use a stamp if you're holding one. Otherwise, it'll
+				// use the clipboard's pen, if applicable.
+				if(!istype(holding, /obj/item/stamp) && clipboard.pen)
+					holding = clipboard.pen
+
+
+			// Safe to assume there are writing implement details as user.can_write(...) fails with an invalid writing implement.
+			var/writing_implement_data = holding.get_writing_implement_details()
+			var/list/field_data = params["field_data"]
+
+			for(var/field_key in field_data)
+				var/field_text = field_data[field_key]
+				var/text_length = length(field_text)
+				if(text_length > MAX_PAPER_INPUT_FIELD_LENGTH)
+					log_paper("[key_name(user)] tried to write to field [field_key] with text over the max limit ([text_length] out of [MAX_PAPER_INPUT_FIELD_LENGTH]) with the following text: [field_text]")
+					return TRUE
+				if(text2num(field_key) >= input_field_count)
+					log_paper("[key_name(user)] tried to write to invalid field [field_key] (when the paper only has [input_field_count] fields) with the following text: [field_text]")
+					return TRUE
+
+				if(!add_field_input(field_key, field_text, writing_implement_data["font"], writing_implement_data["color"], writing_implement_data["use_bold"], user.real_name))
+					log_paper("[key_name(user)] tried to write to field [field_key] when it already has data, with the following text: [field_text]")
+
+			update_static_data(user, ui)
+			return TRUE
+
+/obj/item/paper/proc/get_input_field_count(raw_text)
+	var/static/regex/field_regex = new(@"\[_+\]","g")
+
+	var/counter = 0
+	while(field_regex.Find(raw_text))
+		counter++
+
+	return counter
 
 /obj/item/paper/ui_host(mob/user)
 	if(istype(loc, /obj/structure/noticeboard) || istype(loc, /obj/item/modular_computer))
 		return loc
 	return ..()
 
-/**
- * Construction paper
- */
+/obj/item/paper/proc/get_total_length()
+	var/total_length = 0
+	for(var/datum/paper_input/entry as anything in raw_text_inputs)
+		total_length += length(entry.raw_text)
+
+	return total_length
+
+/// Get a single string representing the text on a page
+/obj/item/paper/proc/get_raw_text()
+	var/paper_contents = ""
+	for(var/datum/paper_input/line as anything in raw_text_inputs)
+		paper_contents += line.raw_text + "/"
+	return paper_contents
+
+/// A single instance of a saved raw input onto paper.
+/datum/paper_input
+	/// Raw, unsanitised, unparsed text for an input.
+	var/raw_text = ""
+	/// Font to draw the input with.
+	var/font = ""
+	/// Colour to draw the input with.
+	var/colour = ""
+	/// Whether to render the font bold or not.
+	var/bold = FALSE
+
+/datum/paper_input/New(_raw_text, _font, _colour, _bold)
+	raw_text = _raw_text
```

Comment:
Raw text should really be sanitised at some point before being stored into a variable (Its any text at all coming from users).

Reply:
Don't know if you could call it sanitisation but it does get sanitised here in the TGUI before it gets saved into the code itself. 
https://github.com/BeeStation/BeeStation-Hornet/blob/1da6cd6a98262468a125cc6699dad76132eade83/tgui/packages/tgui/interfaces/PaperSheet.tsx#L660

Reply:
Ah hmm, so apparently what the code itself is doing is that it's merely saving the raw input, which is fragmented and incomplete. 

According to the Papercode Redux PR by Timberpoes:
> I completely redesigned paper from the ground up to follow a model of composition. It stores individual instances of saved user input, completely raw and unparsed and unsanitised.
<details>
<summary>Large amount of quotes from that PR</summary>

![image](https://user-images.githubusercontent.com/59128051/213993583-311552f0-49b4-4cba-ba0a-71154cc00302.png)

> In my opinion, it isn't even worth worrying about the data stored DM-side, because it can never ever ever be trusted unless we do our own sanitisation and parsing sweep DM side. And we're not wasting our time with that. Because we should only ever be rendering papercode via JS, as spending time making it safe AND renderable outside of tgui is a pointless endeavour.

> There is no single easy and authoritative way to get the raw text out of paper and into someone's browser or chat. It has to be done manully, every time. Because paper input is raw and unsafe and not intended to be used outside of JS.

> All the new raw input components are pieced together into a final result JS-side. tgui is explicitly untrusted - We no longer let it tell us what font to use, what stamp we just put on the document or who the document author is or anything like that. The only user input it doesn't verify are x/y coordinates of stamps - Because the window is arbitrarily resizable and the consequences of trusting where a user put a stamp is minimal-to-no-consequence.

> tgui paper then does its best to match its state with the game based on the ui_data and ui_static_data.

> I don't try saving parsed or sanitised stuff DM-side. I just save the raw input. And the raw input is both fragmented and incomplete. There is only one instance where I ended up copying the raw data, and that was to continue supporting book binding.

</details>

From what I can gather, it stores the raw input by chopping it up into multiple individual instances and only ever lets the complete information be visible to a player after having been forced through the paper TGUI which itself only ever renders things that have been sent through a whitelist found in the sanitize.js file

Found the Coder explaining it in another [PR](https://github.com/tgstation/tgstation/pull/72553#pullrequestreview-1240366940) as well: 
> Paper input is datumised DM-side as `/datum/paper_input`. Every time a user saves modificatgions to paper, a new `/datum/paper_input` is created for it. All input is stored raw DM-side and sanitised TGUI side.

Reply:
Don't know if that's a good enough explanation but I'm pretty sure just telling it to sanitize it from the dm file would break the entire paper code in itself.